File: src/project/registry.ts
import { mkdir, readFile, rename, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';
import { paths } from '../config/paths';

/** A project = a Feishu group bound to a fixed working directory. */
export interface Project {
  /** unique project name (also the group name) */
  name: string;
  /** the bound Feishu group chat_id (oc_xxx) */
  chatId: string;
  /** absolute working directory codex runs in for this project */
  cwd: string;
  /** true when bridge created the cwd as a blank project (under projectsRootDir) */
  blank: boolean;
  createdAt: number;
  /** last branch shown in the announcement (for lazy change detection) */
  branch?: string;
  /** group session model: 'multi' (default) = a topic per session (现状);
   * 'single' = the whole group is one session keyed by chatId. */
  kind?: 'multi' | 'single';
  /** respond to non-@ messages too. Read as `noMention ?? defaultNoMention(p)`.
   * multi: only inside a topic; single: whole group. Needs im:message.group_msg. */
  noMention?: boolean;
  /** how the bot got into this group. 'created' (default, omitted on old data) =
   * bridge built the group via chat.create and is its owner; 'joined' = a human
   * added the bot to a pre-existing group and the bot is just a plain member. */
  origin?: 'created' | 'joined';
  /** for 'joined' projects: open_id of the person who added the bot + did the
   * bind (the bot DMs them the bind card / a removal notice). */
  addedBy?: string;
  /** 项目级响应白名单：谁能让 bot 在本群响应/跑 codex。空/缺省 = 所有人；
   * admin/owner 恒豁免（见 isUserAllowedInProject）。 */
  allowedUsers?: string[];
}

/**
 * Default for 免@ (respond without @) when a project hasn't set `noMention`
 * explicitly. On for everything **except** a *joined* single-session group:
 * making the whole of a pre-existing (possibly busy / multi-person) group run
 * codex on every message without an @ is too aggressive, so that one combo
 * defaults off. Created groups (incl. single) keep the historical default (on),
 * so existing data is unaffected. Single source of truth — every
 * `noMention ?? …` read goes through here.
 */
export function defaultNoMention(p: Pick<Project, 'kind' | 'origin'>): boolean {
  return !((p.origin ?? 'created') === 'joined' && (p.kind ?? 'multi') === 'single');
}

interface StoreFile {
  version: number;
  projects: Project[];
}

const FILE_VERSION = 1;

async function read(): Promise<Project[]> {
  try {
    const text = await readFile(paths.projectsFile, 'utf8');
    const parsed = JSON.parse(text) as Partial<StoreFile>;
    return Array.isArray(parsed.projects) ? parsed.projects : [];
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === 'ENOENT') return [];
    throw err;
  }
}

async function write(projects: Project[]): Promise<void> {
  await mkdir(dirname(paths.projectsFile), { recursive: true });
  const tmp = `${paths.projectsFile}.tmp-${process.pid}`;
  const body: StoreFile = { version: FILE_VERSION, projects };
  await writeFile(tmp, `${JSON.stringify(body, null, 2)}\n`, 'utf8');
  await rename(tmp, paths.projectsFile);
}

export async function listProjects(): Promise<Project[]> {
  return read();
}

export async function getProjectByChatId(chatId: string): Promise<Project | undefined> {
  return (await read()).find((p) => p.chatId === chatId);
}

export async function getProjectByName(name: string): Promise<Project | undefined> {
  return (await read()).find((p) => p.name === name);
}

/** Add a project. Throws if the name — or the bound chat — is already taken.
 * The chatId check is the registry-level hard guard against binding one group
 * twice (createProject's chatId is freshly minted so it never trips). */
export async function addProject(p: Project): Promise<void> {
  const projects = await read();
  if (projects.some((x) => x.name === p.name)) {
    throw new Error(`项目名「${p.name}」已存在`);
  }
  if (p.chatId) {
    const bound = projects.find((x) => x.chatId === p.chatId);
    if (bound) throw new Error(`该群已绑定为项目「${bound.name}」`);
  }
  projects.push(p);
  await write(projects);
}

/** Patch fields of a project by name; no-op if it doesn't exist. */
export async function updateProject(
  name: string,
  patch: Partial<Omit<Project, 'name'>>,
): Promise<void> {
  const projects = await read();
  const p = projects.find((x) => x.name === name);
  if (!p) return;
  const target = p as unknown as Record<string, unknown>;
  for (const [k, v] of Object.entries(patch)) {
    if (v !== undefined) target[k] = v;
  }
  await write(projects);
}

/** Remove (unbind) a project by name. Returns the removed entry, if any. */
export async function removeProject(name: string): Promise<Project | undefined> {
  const projects = await read();
  const idx = projects.findIndex((p) => p.name === name);
  if (idx === -1) return undefined;
  const [removed] = projects.splice(idx, 1);
  await write(projects);
  return removed;
}

File: internal/project/registry.go
package project

// registry.go —— 项目（飞书群↔cwd）注册表（对齐 TS project/registry）。
// 纯数据层：Project 结构 + 纯函数（权限档/免@）+ Store CRUD（并发原子写）。
// 不依赖飞书 SDK（建群/公告等在 lifecycle/onboarding，飞书 SDK 就绪后实现）。

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"

	"github.com/modelzen/feishu-codex-bridge/internal/agent"
)

// Project 一个项目 = 一个飞书群绑定的固定工作目录。
type Project struct {
	Name          string                `json:"name"`   // 唯一项目名（=群名）
	ChatID        string                `json:"chatId"` // 绑定的飞书群 chat_id（oc_xxx）
	Cwd           string                `json:"cwd"`    // codex 工作目录（绝对）
	Blank         bool                  `json:"blank"`  // bridge 在 projectsRootDir 下新建的空项目
	CreatedAt     int64                 `json:"createdAt"`
	Branch        string                `json:"branch,omitempty"`
	Kind          string                `json:"kind,omitempty"`      // multi(默认)|single
	NoMention     *bool                 `json:"noMention,omitempty"` // nil=用 DefaultNoMention
	Origin        string                `json:"origin,omitempty"`    // created(默认)|joined
	AddedBy       string                `json:"addedBy,omitempty"`
	AllowedUsers  []string              `json:"allowedUsers,omitempty"` // 响应白名单；空=所有人
	Mode          agent.PermissionMode  `json:"mode,omitempty"`         // nil/空=full
	GuestMode     agent.PermissionMode  `json:"guestMode,omitempty"`    // 空=同 Mode
	Network       *bool                 `json:"network,omitempty"`
	AutoCompact   *bool                 `json:"autoCompact,omitempty"` // nil=默认 on
	Backend       string                `json:"backend,omitempty"`     // 空=codex 默认
	DefaultModel  string                `json:"defaultModel,omitempty"`
	DefaultEffort agent.ReasoningEffort `json:"defaultEffort,omitempty"`
	SourceURL     string                `json:"sourceUrl,omitempty"`   // 关联的云文档 URL（评论回复定位项目用）
}

// DefaultNoMention 免@默认：除「joined + single」组合外都开。
// （joined 群 + 单会话 = 整个群每条消息都跑 codex，太激进 → 默认关。）
func DefaultNoMention(p Project) bool {
	origin := p.Origin
	if origin == "" {
		origin = "created"
	}
	kind := p.Kind
	if kind == "" {
		kind = "multi"
	}
	return !(origin == "joined" && kind == "single")
}

// EffectiveMode 权限档：空=full（保留历史 danger-full-access）。
func EffectiveMode(p Project) agent.PermissionMode {
	if p.Mode == "" {
		return agent.PermissionFull
	}
	return p.Mode
}

// EffectiveGuestMode 非管理员档：空=同 EffectiveMode（不分级）。
func EffectiveGuestMode(p Project) agent.PermissionMode {
	if p.GuestMode != "" {
		return p.GuestMode
	}
	return EffectiveMode(p)
}

// TurnTier 按 sender 是否 admin 解析权限档 + role + 是否分级。
// split=true 时 admin/guest 跑独立 codex thread（沙箱 + 会话历史都按 role 隔离）。
func TurnTier(p Project, isAdminSender bool) (mode agent.PermissionMode, role string, split bool) {
	admin := EffectiveMode(p)
	guest := EffectiveGuestMode(p)
	if isAdminSender {
		mode, role = admin, "admin"
	} else {
		mode, role = guest, "guest"
	}
	split = guest != admin
	return
}

// ── Store（CRUD + 并发原子写）─────────────────────────────────────

const fileVersion = 1

type storeFile struct {
	Version  int       `json:"version"`
	Projects []Project `json:"projects"`
}

// Store 项目注册表（path 注入；mutex 串行 read-modify-write）。
type Store struct {
	path string
	mu   sync.Mutex
}

// NewStore 构造（path = config.BotProjectsFile(appID)）。
func NewStore(path string) *Store {
	return &Store{path: path}
}

// Path 返回存储文件路径。
func (s *Store) Path() string { return s.path }

func (s *Store) read() ([]Project, error) {
	b, err := os.ReadFile(s.path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var sf storeFile
	if err := json.Unmarshal(b, &sf); err != nil {
		return nil, err
	}
	return sf.Projects, nil
}

func (s *Store) write(projects []Project) error {
	if err := os.MkdirAll(filepath.Dir(s.path), 0o755); err != nil {
		return err
	}
	b, err := json.MarshalIndent(storeFile{Version: fileVersion, Projects: projects}, "", "  ")
	if err != nil {
		return err
	}
	b = append(b, '\n')
	tmp := fmt.Sprintf("%s.tmp-%d", s.path, os.Getpid())
	if err := os.WriteFile(tmp, b, 0o600); err != nil {
		return err
	}
	return os.Rename(tmp, s.path)
}

// List 全部项目。
func (s *Store) List() ([]Project, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.read()
}

// GetByChatID 按群 chat_id 查；未找到返回 (nil, nil)。
func (s *Store) GetByChatID(chatID string) (*Project, error) {
	projects, err := s.List()
	if err != nil {
		return nil, err
	}
	for i := range projects {
		if projects[i].ChatID == chatID {
			return &projects[i], nil
		}
	}
	return nil, nil
}

// GetByName 按项目名查。
func (s *Store) GetByName(name string) (*Project, error) {
	projects, err := s.List()
	if err != nil {
		return nil, err
	}
	for i := range projects {
		if projects[i].Name == name {
			return &projects[i], nil
		}
	}
	return nil, nil
}

// ErrProjectNameExists 项目名已存在。
var ErrProjectNameExists = errors.New("project: name already exists")

// ErrChatAlreadyBound 该群已绑定其它项目。
var ErrChatAlreadyBound = errors.New("project: chat already bound")

// Add 新增项目。name 唯一 + chatId 不重复（注册表级硬守卫，防一群双绑）。
func (s *Store) Add(p Project) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	projects, err := s.read()
	if err != nil {
		return err
	}
	for _, x := range projects {
		if x.Name == p.Name {
			return fmt.Errorf("%w: %s", ErrProjectNameExists, p.Name)
		}
	}
	if p.ChatID != "" {
		for _, x := range projects {
			if x.ChatID == p.ChatID {
				return fmt.Errorf("%w: %s（已绑定 %s）", ErrChatAlreadyBound, p.ChatID, x.Name)
			}
		}
	}
	projects = append(projects, p)
	return s.write(projects)
}

// Update 按名 patch（函数式指针修改，基于最新盘值）；不存在 no-op。
func (s *Store) Update(name string, fn func(*Project)) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	projects, err := s.read()
	if err != nil {
		return err
	}
	for i := range projects {
		if projects[i].Name == name {
			fn(&projects[i])
			return s.write(projects)
		}
	}
	return nil
}

// Remove 按名删除（解绑），返回被删项；不存在返回 (nil, nil)。
func (s *Store) Remove(name string) (*Project, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	projects, err := s.read()
	if err != nil {
		return nil, err
	}
	for i := range projects {
		if projects[i].Name == name {
			removed := projects[i]
			projects = append(projects[:i], projects[i+1:]...)
			if err := s.write(projects); err != nil {
				return nil, err
			}
			return &removed, nil
		}
	}
	return nil, nil
}

// ListProjectsIn 读指定 projects.json（跨 bot 聚合用，ENOENT 返回空）。
func ListProjectsIn(path string) ([]Project, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var sf storeFile
	if err := json.Unmarshal(b, &sf); err != nil {
		return nil, err
	}
	return sf.Projects, nil
}
